README
README, x2 Mark I - Two Axes Platform, Readme file

1. Introduction

2. Mechanics

3. Electronics

Arduino Uno pin designation table
╔════╦═════╦═══════════╦═══════════╦══════════════════════════════════════════╗
║ id ║ pin ║   type    ║ direction ║               description                ║
╠════╬═════╬═══════════╬═══════════╬══════════════════════════════════════════╣
║  0 ║   0 ║ digital   ║ input     ║ Serial communication                     ║
║  1 ║   1 ║ digital   ║ output    ║ Serial communication                     ║
║  2 ║   0 ║ interrupt ║ input     ║ Fan speed sensor                         ║
║  3 ║   3 ║ digital   ║ output    ║ Stepper 1 (activation/speed control)     ║
║  4 ║   4 ║ digital   ║ output    ║ Stepper drive                            ║
║  5 ║   5 ║ digital   ║ output    ║ Stepper 2 (activation/speed control)     ║
║  6 ║   6 ║ digital   ║ output    ║ Stepper 2 (activation/speed control)     ║
║  7 ║   7 ║ digital   ║ output    ║ Stepper drive                            ║
║  8 ║   8 ║ digital   ║ output    ║ Stepper drive                            ║
║  9 ║   9 ║ PWM       ║ output    ║ X2_INTERFACE_D1 (PWM)                    ║
║ 10 ║  10 ║ PWM       ║ output    ║ Fan speed control                        ║
║ 11 ║  11 ║ digital   ║ output    ║ Stepper 1 (activation/speed control)     ║
║ 12 ║  12 ║ digital   ║ output    ║ Stepper drive                            ║
║ 13 ║  13 ║ digital   ║ output    ║ OK LED (status LED)                      ║
║ 14 ║   0 ║ analog    ║ input     ║ Temperature sensor (step motor driver)   ║
║ 15 ║   1 ║ digital   ║ output    ║ X2_INTERFACE_D0 and X2_INTERFACE_A0      ║
║ 16 ║   2 ║ digital   ║ output    ║ Critical LED                             ║
║ 17 ║   3 ║ digital   ║ input     ║ Warning LED                              ║
║ 18 ║   4 ║ digital   ║ output    ║ Power relay                              ║
║ 19 ║   5 ║ digital   ║ input     ║ Power supply DC detection                ║
╚════╩═════╩═══════════╩═══════════╩══════════════════════════════════════════╝

x2 interface pin designation table for Deep viewer
╔════╦═════════════════╦══════════════════════════════════════════════════════╗
║ id ║      pin        ║                  Deep viewer pins                    ║
╠════╬═════════════════╬══════════════════════════════════════════════════════╣
║  1 ║ 1               ║ +5 Vcc                                               ║
║  2 ║ X2_INTERFACE_D1 ║ Ultrasonic sensor echo pin                           ║
║  3 ║ X2_INTERFACE_D0 ║ Ultrasonic sensor trigger pin                        ║
║  4 ║ 4               ║ GND                                                  ║
╚════╩═════════════════╩══════════════════════════════════════════════════════╝

                                      +-----+
         +----[PWR]-------------------| USB |--+
         |                            +-----+  |
         |         GND/RST2  [ ][ ]            |
         |       MOSI2/SCK2  [ ][ ]  A5/SCL[ ] |   C5 
         |          5V/MISO2 [ ][ ]  A4/SDA[ ] |   C4 
         |                             AREF[ ] |
         |                              GND[x] |
         | [ ]N/C                    SCK/13[x] |   B5
         | [ ]v.ref                 MISO/12[x] |   .
         | [ ]RST                   MOSI/11[x]~|   .
         | [ ]3V3    +---+               10[x]~|   .
         | [x]5v     | A |                9[x]~|   .
         | [x]GND   -| R |-               8[x] |   B0
         | [x]GND   -| D |-                    |
         | [x]Vin   -| U |-               7[x] |   D7
         |          -| I |-               6[x]~|   .
         | [x]A0    -| N |-               5[x]~|   .
         | [x]A1    -| O |-               4[x] |   .
         | [x]A2     +---+           INT1/3[x]~|   .
         | [x]A3                     INT0/2[x] |   .
         | [x]A4/SDA  RST SCK MISO     TX>1[x] |   .
         | [x]A5/SCL  [ ] [ ] [ ]      RX<0[x] |   D0
         |            [ ] [ ] [ ]              |
         |  UNO_R3    GND MOSI 5V  ____________/
          \_______________________/
                                Source: http://busyducks.com/ascii-art-arduinos

4. Software

4.1. Odd Scanner

Use oddscanner program to generate a G-code file:

  oddscanner 100 80 > example.gcode
  
Where:

  oddscanner: program file
  100: width
  80: height
  example.gcode: output G-code file
  
After file generation, use xC to run G-code file:

  xc cli --id x2 -p example.gcode -o example.csv
  
A CSV file will be generated. So this file can be converted to a picture.


CHANGELOG
CHANGELOG, x2 Mark I - Two Axes Platform, Change log file

2017-05-15
        * Improvement: Added power detection before any move command.

2017-04-30
        * Improvement: More documentation.

2017-03-13
        * Bug fix: Command "G1 X0 Y0" was not working.

2017-03-10
        * Bug fix: Odd Scanner outputing wrong picture dimensions.

2017-03-09
        * New feature: Added program to generate Odd Scanner G-code file to be
          used with xC.

2017-03-06
        * Bug fix: Fixed G1 command to move between hypotenuse points.

2017-02-28
        * New feature: command M0.
        * New feature: command M17.
        * New feature: command M18.

2015-09-05
        * New feature: command_laser.

2015-06-07
        * New feature: standby_handler.

2015-01-21
        * New feature: command_power.

2015-01-09
        * Code optimization: I separated the main sketch file into 2 new ones
          to maintain code (handlers.ino and commands.ino).

2014-12-02
        * New feature: CommandG3.

2014-11-27
        * New feature: CommandG6.

2014-11-22
        * Functional version.

201x-xx-xx
        * Grow a lot...

2012-08-11
        * Experimental firmware version.

2006-xx-xx
        * Hardware is almost done.


x2/SCommand.h
/******************************************************************************* 
SerialCommand - An Arduino library to tokenize and parse commands received over
a serial port. 
Copyright (C) 2011-2013 Steven Cogswell  <steven.cogswell@gmail.com>
http://awtfy.com

Version 20131021A.   

Version History:
May 11 2011 - Initial version
May 13 2011 -	Prevent overwriting bounds of SerialCommandCallback[] array in addCommand()
			defaultHandler() for non-matching commands
Mar 2012 - Some const char * changes to make compiler happier about deprecated warnings.  
           Arduino 1.0 compatibility (Arduino.h header) 
Oct 2013 - SerialCommand object can be created using a SoftwareSerial object, for SoftwareSerial
           support.  Requires #include <SoftwareSerial.h> in your sketch even if you don't use 
           a SoftwareSerial port in the project.  sigh.   See Example Sketch for usage. 
Oct 2013 - Conditional compilation for the SoftwareSerial support, in case you really, really
           hate it and want it removed.  

This library is free software; you can redistribute it and/or
modify it under the terms of the GNU Lesser General Public
License as published by the Free Software Foundation; either
version 2.1 of the License, or (at your option) any later version.

This library is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public
License along with this library; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
***********************************************************************************/
#ifndef SerialCommand_h
#define SerialCommand_h

#if defined(ARDUINO) && ARDUINO >= 100
#include "Arduino.h"
#else
#include "WProgram.h"
#endif

// If you want to use SerialCommand with the hardware serial port only, and want to disable
// SoftwareSerial support, and thus don't have to use "#include <SoftwareSerial.h>" in your
// sketches, then uncomment this define for SERIALCOMMAND_HARDWAREONLY, and comment out the 
// corresponding #undef line.  
//
// You don't have to use SoftwareSerial features if this is not defined, you can still only use 
// the Hardware serial port, just that this way lets you get out of having to include 
// the SoftwareSerial.h header. 
#define SERIALCOMMAND_HARDWAREONLY 1
//#undef SERIALCOMMAND_HARDWAREONLY

#ifdef SERIALCOMMAND_HARDWAREONLY
#warning "Warning: Building SerialCommand without SoftwareSerial Support"
#endif

#ifndef SERIALCOMMAND_HARDWAREONLY 
#include <SoftwareSerial.h>  
#endif

#include <string.h>

#define SERIALCOMMANDBUFFER 11
#define MAXSERIALCOMMANDS	20
#define MAXDELIMETER 2

#define SERIALCOMMANDDEBUG 1
#undef SERIALCOMMANDDEBUG      // Comment this out to run the library in debug mode (verbose messages)

class SerialCommand
{
	public:
		SerialCommand();      // Constructor
		#ifndef SERIALCOMMAND_HARDWAREONLY
		SerialCommand(SoftwareSerial &SoftSer);  // Constructor for using SoftwareSerial objects
		#endif

		void clearBuffer();   // Sets the command buffer to all '\0' (nulls)
		char *next();         // returns pointer to next token found in command buffer (for getting arguments to commands)
		void readSerial();    // Main entry point.  
		void addCommand(const char *, void(*)());   // Add commands to processing dictionary
		void addDefaultHandler(void (*function)());    // A handler to call when no valid command received. 
	
	private:
		char inChar;          // A character read from the serial stream 
		char buffer[SERIALCOMMANDBUFFER];   // Buffer of stored characters while waiting for terminator character
		int  bufPos;                        // Current position in the buffer
		char delim[MAXDELIMETER];           // null-terminated list of character to be used as delimeters for tokenizing (default " ")
		char term;                          // Character that signals end of command (default '\r')
		char *token;                        // Returned token from the command buffer as returned by strtok_r
		char *last;                         // State variable used by strtok_r during processing
		typedef struct _callback {
			char command[SERIALCOMMANDBUFFER];
			void (*function)();
		} SerialCommandCallback;            // Data structure to hold Command/Handler function key-value pairs
		int numCommand;
		SerialCommandCallback CommandList[MAXSERIALCOMMANDS];   // Actual definition for command/handler array
		void (*defaultHandler)();           // Pointer to the default handler function 
		int usingSoftwareSerial;            // Used as boolean to see if we're using SoftwareSerial object or not
		#ifndef SERIALCOMMAND_HARDWAREONLY 
		SoftwareSerial *SoftSerial;         // Pointer to a user-created SoftwareSerial object
		#endif
};

#endif //SerialCommand_h


x2/SCommand.cpp
/******************************************************************************* 
SerialCommand - An Arduino library to tokenize and parse commands received over
a serial port. 
Copyright (C) 2011-2013 Steven Cogswell  <steven.cogswell@gmail.com>
http://awtfy.com

See SerialCommand.h for version history. 

This library is free software; you can redistribute it and/or
modify it under the terms of the GNU Lesser General Public
License as published by the Free Software Foundation; either
version 2.1 of the License, or (at your option) any later version.

This library is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public
License along with this library; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
***********************************************************************************/

#if defined(ARDUINO) && ARDUINO >= 100
#include "Arduino.h"
#else
#include "WProgram.h"
#endif

#include "SCommand.h"


#include <string.h>
#ifndef SERIALCOMMAND_HARDWAREONLY
#include <SoftwareSerial.h>
#endif

// Constructor makes sure some things are set. 
SerialCommand::SerialCommand()
{
	usingSoftwareSerial=0;
	strncpy(delim," ",MAXDELIMETER);  // strtok_r needs a null-terminated string
	term='\r';   // return character, default terminator for commands
	numCommand=0;    // Number of callback handlers installed
	clearBuffer(); 
}

#ifndef SERIALCOMMAND_HARDWAREONLY
// Constructor to use a SoftwareSerial object
SerialCommand::SerialCommand(SoftwareSerial &_SoftSer)
{
	usingSoftwareSerial=1; 
	SoftSerial = &_SoftSer;
	strncpy(delim," ",MAXDELIMETER);  // strtok_r needs a null-terminated string
	term='\r';   // return character, default terminator for commands
	numCommand=0;    // Number of callback handlers installed
	clearBuffer(); 
}
#endif


//
// Initialize the command buffer being processed to all null characters
//
void SerialCommand::clearBuffer()
{
	for (int i=0; i<SERIALCOMMANDBUFFER; i++) 
	{
		buffer[i]='\0';
	}
	bufPos=0; 
}

// Retrieve the next token ("word" or "argument") from the Command buffer.  
// returns a NULL if no more tokens exist.   
char *SerialCommand::next() 
{
	char *nextToken;
	nextToken = strtok_r(NULL, delim, &last); 
	return nextToken; 
}

// This checks the Serial stream for characters, and assembles them into a buffer.  
// When the terminator character (default '\r') is seen, it starts parsing the 
// buffer for a prefix command, and calls handlers setup by addCommand() member
void SerialCommand::readSerial() 
{
	// If we're using the Hardware port, check it.   Otherwise check the user-created SoftwareSerial Port
	#ifdef SERIALCOMMAND_HARDWAREONLY
	while (Serial.available() > 0) 
	#else
	while ((usingSoftwareSerial==0 && Serial.available() > 0) || (usingSoftwareSerial==1 && SoftSerial->available() > 0) )
	#endif
	{
		int i; 
		boolean matched; 
		if (usingSoftwareSerial==0) {
			// Hardware serial port
			inChar=Serial.read();   // Read single available character, there may be more waiting
		} else {
			#ifndef SERIALCOMMAND_HARDWAREONLY
			// SoftwareSerial port
			inChar = SoftSerial->read();   // Read single available character, there may be more waiting
			#endif
		}
		#ifdef SERIALCOMMANDDEBUG
		Serial.print(inChar);   // Echo back to serial stream
		#endif
		if (inChar==term) {     // Check for the terminator (default '\r') meaning end of command
			#ifdef SERIALCOMMANDDEBUG
			Serial.print("Received: "); 
			Serial.println(buffer);
		    #endif
			bufPos=0;           // Reset to start of buffer
			token = strtok_r(buffer,delim,&last);   // Search for command at start of buffer
			if (token == NULL) return; 
			matched=false; 
			for (i=0; i<numCommand; i++) {
				#ifdef SERIALCOMMANDDEBUG
				Serial.print("Comparing ["); 
				Serial.print(token); 
				Serial.print("] to [");
				Serial.print(CommandList[i].command);
				Serial.println("]");
				#endif
				// Compare the found command against the list of known commands for a match
				if (strncmp(token,CommandList[i].command,SERIALCOMMANDBUFFER) == 0) 
				{
					#ifdef SERIALCOMMANDDEBUG
					Serial.print("Matched Command: "); 
					Serial.println(token);
					#endif
					// Execute the stored handler function for the command
					(*CommandList[i].function)(); 
					clearBuffer(); 
					matched=true; 
					break; 
				}
			}
			if (matched==false) {
				(*defaultHandler)(); 
				clearBuffer(); 
			}

		}
		if (isprint(inChar))   // Only printable characters into the buffer
		{
			buffer[bufPos++]=inChar;   // Put character into buffer
			buffer[bufPos]='\0';  // Null terminate
			if (bufPos > SERIALCOMMANDBUFFER-1) bufPos=0; // wrap buffer around if full  
		}
	}
}

// Adds a "command" and a handler function to the list of available commands.  
// This is used for matching a found token in the buffer, and gives the pointer
// to the handler function to deal with it. 
void SerialCommand::addCommand(const char *command, void (*function)())
{
	if (numCommand < MAXSERIALCOMMANDS) {
		#ifdef SERIALCOMMANDDEBUG
		Serial.print(numCommand); 
		Serial.print("-"); 
		Serial.print("Adding command for "); 
		Serial.println(command); 
		#endif
		
		strncpy(CommandList[numCommand].command,command,SERIALCOMMANDBUFFER); 
		CommandList[numCommand].function = function; 
		numCommand++; 
	} else {
		// In this case, you tried to push more commands into the buffer than it is compiled to hold.  
		// Not much we can do since there is no real visible error assertion, we just ignore adding
		// the command
		#ifdef SERIALCOMMANDDEBUG
		Serial.println("Too many handlers - recompile changing MAXSERIALCOMMANDS"); 
		#endif 
	}
}

// This sets up a handler to be called in the event that the receveived command string
// isn't in the list of things with handlers.
void SerialCommand::addDefaultHandler(void (*function)())
{
	defaultHandler = function;
}


x2/config.h
/* config.h, x2 Mark I - Two Axes Platform, Arduino project config file
 * 
 * Author: Márcio Pessoa <marcio.pessoa@sciemon.com>
 * Contributors: none
 */

// x2 dedicated hardware user interface
const byte X2_INTERFACE_A0 = 1;
const byte X2_INTERFACE_D0 = 15;
const byte X2_INTERFACE_D1 = 9;

// LED
const byte led_ok_pin = 13;
const byte led_warning_pin = 17;
const byte led_critical_pin = 16;

// Power
const byte power_relay_pin = 18;
const byte power_sensor_pin = 19;

// Fan
const byte fan_control_pin = 10;
const byte fan_sensor_pin = 0;

// Temperature
const byte lm35_pin = 0;

// Ultrasonic distance sensor
const byte ultrasonic_trigger_pin = X2_INTERFACE_D0;
const byte ultrasonic_echo_pin = X2_INTERFACE_D1;

// Laser
const byte laser_pin = X2_INTERFACE_D0;

// Random number generator Seed pin
const byte random_Seed_pin = lm35_pin;

// Timers
const byte health_check_timer = 2;  // seconds
const byte sensors_timer = 2;  // seconds
const byte standby_timer = 1;  // minute

// System status (initial state)
byte general_status = UNKNOWN;

// Debug mode
bool debug_mode = false;

// Serial speed: 4800, 9600, 14400, 19200, 28800, 38400, 57600, 115200
const unsigned long serial_speed = 115200;

// Move done
bool done = true;

// 
const float FLIMIT = 340282350000000000000000000000000000000;
